File: web/app/model/AnalysisModels/QualityReport.js
/*
 * (C) Copyright 2014 The Genomics of Gene Expression Lab, CIPF 
 * (http://bioinfo.cipf.es/aconesawp) and others.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Lesser General Public License
 * (LGPL) version 3 which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/lgpl.html
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * Contributors:
 *     Rafael Hernandez de Diego, rhernandez@cipf.es
 *     Ana Conesa Cegarra, aconesa@cipf.es
 *     
 * THIS FILE CONTAINS THE FOLLOWING COMPONENT DECLARATION
 * - QualityReport
 */
Ext.define('SL.model.AnalysisModels.QualityReport', {
    extend: 'Ext.data.Model',
    alias: 'widget.QualityReport',
    //Extends the Observable class
    mixins: {Observable: 'SL.view.senchaExtensions.Observable'},
    fields: [
        {name: 'studied_step_id'},
        {name: 'software'},
        {name: 'software_version'},
        {name: 'software_configuration'},
        {name: 'results'},
        {name: 'files_location'},
        {name: 'submission_date', convert: function(v, rec) {
                return rec.dateConversion(v);
            }}
    ],
    /**BC********************************************************************
     **
     **GETTERS AND SETTERS
     **
     *EC********************************************************************/
    getStudiedStepID: function() {
        return this.get('studied_step_id');
    },
    setStudiedStepID: function(studied_step_id) {
        this.set('studied_step_id', studied_step_id);
        this.setChanged();
    },
    getSoftware: function() {
        return this.get('software');
    },
    setSoftware: function(software) {
        this.set('software', software);
        this.setChanged();
    },
    getSoftwareVersion: function() {
        return this.get('software_version');
    },
    setSoftwareVersion: function(software_version) {
        this.set('software_version', software_version);
        this.setChanged();
    },
    getSoftwareConfiguration: function() {
        return this.get('software_configuration');
    },
    setSoftwareConfiguration: function(software_configuration) {
        this.set('software_configuration', software_configuration);
        this.setChanged();
    },
    getResults: function() {
        return this.get('results');
    },
    setResults: function(results) {
        this.set('results', results);
        this.setChanged();
    },
    getFileLocation: function() {
        return this.get('files_location');
    },
    setFileLocation: function(files_location) {
        this.set('files_location', files_location);
        this.setChanged();
    },
    getSubmissionDate: function() {
        return this.get('submission_date');
    },
    setSubmissionDate: function(submission_date) {
        this.set('submission_date', submission_date);
        this.setChanged();
    },
    isEmpty: function() {
        return (this.getSoftware() === "" || this.getSoftware() == null) &&
                this.getSoftwareVersion() === "" &&
                (this.getSoftwareConfiguration() === "" || this.getSoftwareConfiguration() == null) &&
                this.getResults() === "" &&
                this.getSubmissionDate() === "";
    },
    statics: {
        /********************************************************************************      
         * This static function creates a new Intermediate step MODEL using the information 
         * provided as a JSON
         *  
         * @param jsonData, the data for the Intermediate step model in JSON format
         * @return the model with the loaded data    
         ********************************************************************************/
        loadFromJSON: function(jsonData) {
            var model = Ext.create('SL.model.AnalysisModels.QualityReport');
            model.set(jsonData);
            return model;
        }
    },
    getJSONforGraph: function() {
        var json_data = {
            id: "qr_" + this.getStudiedStepID(),
            name: "",
            type: "Quality report",
            nodeType: "Quality_report",
            summary: "TODO",
            parents_id: [this.getStudiedStepID()],
        };
        return json_data;
    },
    toSimpleJSON: function() {
        return this.data;
    },
    dateConversion: function(v) {
        if (v == null || v.length < 8) {
            return ""
        }
        if (v instanceof Date) {
            var formatNumber = function(number, length) {
                var str = '' + number;
                while (str.length < length) {
                    str = '0' + str;
                }
                return str;
            }
            v = v.getFullYear() + "/" + (formatNumber(v.getMonth() + 1, 2)) + "/" + formatNumber(v.getDate(), 2);
            return v;
        }
        if (v.indexOf("/") === -1)
            return v.slice(0, 4) + "/" + v.slice(4, 6) + "/" + v.slice(6, 8);
        else
            return v;
    }

});

File: src/java/classes/analysis/QualityReport.java
/* ***************************************************************
 *  This file is part of STATegra EMS.
 *
 *  STATegra EMS is free software: you can redistribute it and/or 
 *  modify it under the terms of the GNU General Public License as
 *  published by the Free Software Foundation, either version 3 of 
 *  the License, or (at your option) any later version.
 * 
 *  STATegra EMS is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with STATegra EMS.  If not, see <http://www.gnu.org/licenses/>.
 * 
 *  More info http://bioinfo.cipf.es/stategraems
 *  Technical contact stategraemsdev@gmail.com
 *  *************************************************************** */

package classes.analysis;

import com.google.gson.Gson;

/**
 *
 * @author Rafa Hernández de Diego
 */
public class QualityReport {

    private String studied_step_id;
    private String software;
    private String software_version;
    private String software_configuration;
    private String results;
    private String files_location;
    private String submission_date;

    public QualityReport(String studied_step_id, String software, String software_version, String software_configuration, String results, String files_location,String submission_date) {
        this.studied_step_id = studied_step_id;
        this.software = software;
        this.software_version = software_version;
        this.software_configuration = software_configuration;
        this.results = results;
        this.files_location = files_location;
        this.submission_date = submission_date;
    }
    
    /**
     * This static function returns a new object using the data contained in the
     * given JSON object (as String).
     *
     * @param jsonString the JSON object
     * @return the new Object.
     */
    public static QualityReport fromJSON(String jsonString) {
        Gson gson = new Gson();
        QualityReport quality_report = gson.fromJson(jsonString, QualityReport.class);

        return quality_report;
    }

    public String toJSON() {
        Gson gson = new Gson();
        String jsonString = gson.toJson(this);

        return jsonString;
    }

    
    //**********************************************************************
    //* GETTERS AND SETTERS ************************************************
    //**********************************************************************
    
    public String getStudied_step_id() {
        return studied_step_id;
    }

    public void setStudiedStepID(String studied_step_id) {
        this.studied_step_id = studied_step_id;
    }

    public String getSoftware() {
        return software;
    }

    public void setSoftware(String software) {
        this.software = software;
    }

    public String getSoftware_version() {
        return software_version;
    }

    public void setSoftware_version(String software_version) {
        this.software_version = software_version;
    }

    public String getResults() {
        return results;
    }

    public void setResults(String results) {
        this.results = results;
    }

    public String getFiles_location() {
        return files_location;
    }

    public void setFiles_location(String files_location) {
        this.files_location = files_location;
    }

    public String getSoftware_configuration() {
        return software_configuration;
    }

    public void setSoftware_configuration(String software_configuration) {
        this.software_configuration = software_configuration;
    }

    public String getSubmissionDate() {
        return submission_date;
    }

    public void setSubmissionDate(String submission_date) {
        this.submission_date = submission_date;
    }

    
    //***********************************************************************
    //* OTHER FUNCTIONS *****************************************************
    //***********************************************************************
    
    @Override
    public String toString() {
        return this.toJSON();
    }
    
}
